src/Domain/FunctionsOfTime/QuaternionHelpers.cpp

```diff
@@ -32,5 +34,8 @@ boost::math::quaternion<double> datavector_to_quaternion(
 // Normalize a `boost::math::quaternion`
 void normalize_quaternion(
     const gsl::not_null<boost::math::quaternion<double>*> input) noexcept {
+  if (equal_within_roundoff(abs(*input), 0.0)) {
+    ERROR("Vanishing quaternion!");
+  }
```

Comment:
```suggestions
  ASSERT(not equal_within_roundoff(abs(*input), "Vanishing quaternion!");
```

since this is a programmer error. User-facing errors would need some more context on what might be going wrong, e.g. in the control system.

Also, does this actually improve the error over the raw FPE that would be thrown anyway?

Reply:
I specifically made this a run-time error because in my experience with SpEC, sometimes when a simulation goes bad the quaternion just vanishes. Thus, in a release, the ASSERT wouldn't catch this.

And yes the FPE isn't very descriptive so this error message at least narrows down *why* there was a FPE and makes it easier to trace back (as usually something other than the quaternion is the root cause of a vanishing quaternion)

Reply:
Hm, I think the error-checking for a user-facing error should be done earlier in the call stack where you have more context. At this point I see no difference between "FPE in normalize_quaternion" and "ERROR: Vanishing quaternion! in normalize_quaternion".

Reply:
Error-checking earlier in the stack would be better but as of right now I'm unclear on how that would look.

Either way, I'll change this to an ASSERT because if I remember correctly, the reason I added this was because I got a FPE while testing but it didn't say from where so it took a while to track down.

Reply:
I think it's better to just drop the change, or is there any information the assert/error adds to the FPE?

If you want to add better errors, I suggest you do so in `QuaternionFunctionOfTime::update_stored_info()` and `QuaternionFunctionOfTime::setup_func()`. There you have the context available to output useful information.

Reply:
Oh and when you get an FPE, a very effective way to track it down is to run the executable in a debugger, like `gdb`, ideally in an IDE that jumps directly to the offending line in the source code. In fact, I'm always happy when I get an FPE because those are relatively easy to find :)

Reply:
I'll drop the change.